### PR TITLE
Added test cases for A-C rules (C#11)

### PR DESCRIPTION
**Pull request by cristian-ambrosini-sonarsource**

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/CompareNaN.CSharp11.cs [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    void ListPattern()
+    {
+        object[] numbers = { 1, 2, double.NaN };
+
+        if (numbers is [1, 2, double.NaN]) // FN
+        {
+            Console.WriteLine("Test1");
+        }
+
+        if (numbers is [1, double.NaN, 3]) // FN
```
Comment (csaba-sagi-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/CompareNaN.CSharp11.cs:15]: Why do you have this second test case? Why is it different from the first one?

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ControlCharacterInString.CSharp11.cs [hunk 2]
```diff
@@ -4,7 +4,18 @@ namespace Tests.Diagnostics
 {
     class Program
     {
-        public const string Part1 = """test"""; // Noncompliant
-        public const string Part2 = """test""";
+        public const string RawStringLiteralsNonCompliant = """test"""; // Noncompliant
+        public const string RawStringLiteralsCompliant = """test""";
+
+        public const string InterpolatedStringNonCompliant = $"""test{RawStringLiteralsNonCompliant}"""; // FN
+        public const string InterpolatedStringCompliant = $"""test{RawStringLiteralsCompliant}""";
```
Comment (csaba-sagi-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ControlCharacterInString.CSharp11.cs:11]: Besides this case, there should be also one where the control character is in the text part of the interpolated string and not the interpolation.

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/BinaryOperationWithIdenticalExpressionsTest.cs [hunk 3]
```diff
@@ -26,26 +26,39 @@ namespace SonarAnalyzer.UnitTest.Rules
...
-                new CS.BinaryOperationWithIdenticalExpressions());
+            builderCS.AddPaths("BinaryOperationWithIdenticalExpressions.cs").Verify();
 
         [TestMethod]
         public void BinaryOperationWithIdenticalExpressions_TestProject_CS() =>
-            OldVerifier.VerifyNoIssueReported(@"TestCases\BinaryOperationWithIdenticalExpressions.cs",
-                new CS.BinaryOperationWithIdenticalExpressions(),
-                TestHelper.ProjectTypeReference(SonarAnalyzer.Helpers.ProjectType.Test));
+            builderCS.AddPaths("BinaryOperationWithIdenticalExpressions.cs")
+                .AddReferences(TestHelper.ProjectTypeReference(ProjectType.Test))
```
Comment (csaba-sagi-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Rules/BinaryOperationWithIdenticalExpressionsTest.cs:39]: Instead of the old way of adding test reference, you should use:
```suggestion
                .AddTestReference()
```

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/BinaryOperationWithIdenticalExpressionsTest.cs [hunk 4]
```diff
@@ -26,26 +26,39 @@ namespace SonarAnalyzer.UnitTest.Rules
...
-                new VB.BinaryOperationWithIdenticalExpressions());
+            builderVB.AddPaths("BinaryOperationWithIdenticalExpressions.vb").Verify();
 
         [TestMethod]
         public void BinaryOperationWithIdenticalExpressions_TestProject_VB() =>
-            OldVerifier.VerifyNoIssueReported(@"TestCases\BinaryOperationWithIdenticalExpressions.vb",
-                new VB.BinaryOperationWithIdenticalExpressions(),
-                TestHelper.ProjectTypeReference(SonarAnalyzer.Helpers.ProjectType.Test));
+            builderVB.AddPaths("BinaryOperationWithIdenticalExpressions.vb")
+                .AddReferences(TestHelper.ProjectTypeReference(ProjectType.Test))
```
Comment (csaba-sagi-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Rules/BinaryOperationWithIdenticalExpressionsTest.cs:49]: Applies here too.

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/BinaryOperationWithIdenticalExpressionsTest.cs [hunk 5]
```diff
@@ -26,26 +26,39 @@ namespace SonarAnalyzer.UnitTest.Rules
...
+                .VerifyNoIssueReported();
+
+#if NET
+
+        [TestMethod]
+        public void BinaryOperationWithIdenticalExpressions_CSharp11() =>
+            builderCS.AddPaths("BinaryOperationWithIdenticalExpressions.CSharp11.cs")
+                .WithTopLevelStatements()
```
Comment (csaba-sagi-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Rules/BinaryOperationWithIdenticalExpressionsTest.cs:57]: You can move the `UnsignedRightShiftOperator` and `OverloadableOperators` local functions to be methods of a class, and you will not need o use top level statements here.

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/CastShouldNotBeDuplicatedTest.cs [hunk 6]
```diff
@@ -25,18 +25,34 @@ namespace SonarAnalyzer.UnitTest.Rules
...
         public void CastShouldNotBeDuplicated_CSharp10() =>
-            OldVerifier.VerifyAnalyzerFromCSharp10Library(@"TestCases\CastShouldNotBeDuplicated.CSharp10.cs", new CastShouldNotBeDuplicated());
+            builder.AddPaths("CastShouldNotBeDuplicated.CSharp10.cs")
+                .WithOptions(ParseOptionsHelper.FromCSharp10)
+                .Verify();
+
+        [TestMethod]
+        public void CastShouldNotBeDuplicated_CSharp11() =>
+            builder.AddPaths("CastShouldNotBeDuplicated.CSharp11.cs")
+                .WithTopLevelStatements()
```
Comment (csaba-sagi-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Rules/CastShouldNotBeDuplicatedTest.cs:51]: void ListPattern() can go into a class and you will not need the WithTopLevelStatements.

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/CognitiveComplexityTest.cs [hunk 7]
```diff
@@ -33,29 +33,37 @@ public class CognitiveComplexityTest
...
-            builderCS.AddPaths(@"CognitiveComplexity.CSharp9.cs")
+            builderCS.AddPaths("CognitiveComplexity.CSharp9.cs")
                 .WithTopLevelStatements()
                 .Verify();
 
         [TestMethod]
         public void CognitiveComplexity_CS_CSharp10() =>
-            builderCS.AddPaths(@"CognitiveComplexity.CSharp10.cs")
+            builderCS.AddPaths("CognitiveComplexity.CSharp10.cs")
                 .WithLanguageVersion(LanguageVersion.CSharp10)
```
Comment (csaba-sagi-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Rules/CognitiveComplexityTest.cs:57]: Better to use .WithOptions(ParseOptionsHelper.FromCSharp10) instead of WithLanguageVersion. Locally there is no difference but on the CI, FromChsarp10 will also run it for CSHarp11.

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/CognitiveComplexityTest.cs [hunk 8]
```diff
@@ -33,29 +33,37 @@ public class CognitiveComplexityTest
...
                 .WithConcurrentAnalysis(false)
                 .Verify();
+
+        [TestMethod]
+        public void CognitiveComplexity_CS_CSharp11() =>
+            builderCS.AddPaths("CognitiveComplexity.CSharp11.cs")
+                .WithLanguageVersion(LanguageVersion.CSharp11)
```
Comment (csaba-sagi-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Rules/CognitiveComplexityTest.cs:64]: Here too.

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/ConstructorOverridableCallTest.cs [hunk 9]
```diff
@@ -25,8 +25,12 @@ namespace SonarAnalyzer.UnitTest.Rules
     [TestClass]
     public class ConstructorOverridableCallTest
     {
+        private readonly VerifierBuilder builder = new VerifierBuilder<ConstructorOverridableCall>();
+
         [TestMethod]
         public void ConstructorOverridableCall() =>
-            OldVerifier.VerifyAnalyzer(@"TestCases\ConstructorOverridableCall.cs", new ConstructorOverridableCall());
+            builder.AddPaths("ConstructorOverridableCall.cs")
+                .WithConcurrentAnalysis(false)
```
Comment (csaba-sagi-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Rules/ConstructorOverridableCallTest.cs:33]: Why did you add here WithConcurrentAnalysis(false)?

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ConstructorOverridableCall.cs [hunk 10]
```diff
@@ -55,4 +55,32 @@ public override void DoSomething()
...
+        public string Name { get; set; }
+
+        public Class3(string name)
+        {
+            Name = name;
+        }
+
+        public override void DoSomething()
+        {
+            base.DoSomething();
```
Comment (csaba-sagi-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ConstructorOverridableCall.cs:83]: To really show what the problem is here, instead of calling base.DoSomething, you should do something with Name property as it should be null at this point.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ConstructorOverridableCall.cs [hunk 11]
```diff
@@ -55,4 +55,32 @@ public override void DoSomething()
...
+    public class Class1
+    {
+        public virtual void DoSomething() { }
+    }
+
+    public class Class2 : Class1
+    {
+        public Class2()
+        {
+            DoSomething(); // FN
```
Comment (csaba-sagi-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ConstructorOverridableCall.cs:68]: Also please don't forget to open an issue for this FN as it is not C#11 related.

**Review by csaba-sagi-sonarsource — CHANGES_REQUESTED** [review]
One more round of changes requested.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/CompareNaN.CSharp11.cs (hunk 1 above)
Comment (cristian-ambrosini-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/CompareNaN.CSharp11.cs:15]: The first one is comparing NaN to NaN, and the second Int32 to Nan, I thought that doing that should cover more expressively the pattern. Which one do you think should be removed?

Comment (csaba-sagi-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/CompareNaN.CSharp11.cs]: Discussed offline.
